appvia/kore
# Introduce IP filtering for the Kubernetes API + the auth proxy (separately)

## What

Story: https://github.com/appvia/kore-planning/issues/114

This adds IP filtering for the Kubernetes API and for the auth proxy.

I've separated the original `enablePrivateNetwork` plan option into two:
 * `enablePrivateNetwork`: this makes the Kubernetes nodes private (so no external IP)
 * `enablePrivateEndpoint`: this makes the Kubernetes API private (access from internal IP only)

The `authorizedMasterNetworks` option can now be set independently from the above options and can be used do enable IP filtering for the Kubernetes API.

The PR adds the `allowed-ips` option to auth-proxy, to allow filtering incoming requests. I had to set `externalTrafficPolicy` to `Local` on the auth proxy service, so the proxy gets the real client IP instead of a local one.

I've introduced some tests for the auth proxy which required a mocking library (counterfeiter), http testing library (gomega's ghttp) and the refactoring of the auth proxy + some new interfaces.

## Additional changes

 * Log unknown cloud provider warning in the correct place
 * When creating a cluster with --dry-run, do not wait for the creation
 * Make the idp-client-id option for auth proxy required as it is mandatory
 * Allow to override which Docker images to build/push
 * Remove unused gomega/gbytes and gomega/gexec dependencies

## Related PRs

 * Allow to override plan parameters in the CLI: https://github.com/appvia/kore/pull/335

## How to test

 * Build and push a custom auth proxy image
    ```
    DOCKER_IMAGES="auth-proxy" VERSION=private_gcp_cluster make docker-release
    ```
 * Start the API with setting the custom Docker image
    ```
    AUTH_PROXY_IMAGE=quay.io/appvia/auth-proxy:private_gcp_cluster make run
    ```
 * Using https://github.com/appvia/kore/pull/361 create a cluster with your own IP address:
    ```
    korectl create cluster \
    --team a-team andras-test \
    --plan gke-development \
    --allocation gke \
    --plan-param authorizedMasterNetworks='[{"name":"home", "cidr": "1.2.3.4/32"}]' \
    --plan-param authProxyAllowedIPs='["1.2.3.4/32"]'
    ```
 * Auth to the cluster:
    ```
    korectl clusters auth -n andras-test --team a-team
    kubectl config use-context andras-test
    ```
 * Make sure you can connect to the K8s API through the auth proxy:
    ```
    kubectl get namespaces
    ```
 * Make sure you can still access the K8s API directly. Get caCertificate and apiEndpoint from
    ```
    korectl get cluster -t a-team andras-test -o yaml
    ```
    Curl the API directly, it should have a JSON response:
    ```
    curl --cacert ca.cert https://<apiEndpoint>
    ```
 * Hop on a VPN if you have one and make sure you are not able to access the auth-proxy or the K8s API anymore
 * Delete the cluster
    ```
    korectl delete cluster andras-test --team a-team
    ```
